consensus/view_change.go

```diff
@@ -90,20 +90,111 @@ func (pm *State) GetViewChangeDuraion() time.Duration {
+// The next view ID is the single factor used to determine
+// the next leader, so it is mod the number of nodes per shard.
+// It returns the next viewID and duration of the view change
+// The view change duration is a fixed duration now to avoid stuck into offline nodes during
+// the view change.
+// viewID is only used as the fallback mechansim to determine the nextViewID
+func (consensus *Consensus) getNextViewID(viewID uint64) (uint64, time.Duration) {
+	// handle corner case at first
+	if consensus.ChainReader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	curHeader := consensus.ChainReader.CurrentHeader()
+	if curHeader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	blockTimestamp := curHeader.Time().Int64()
+	curTimestamp := time.Now().Unix()
+
+	// timestamp messed up in current validator node
+	if curTimestamp < blockTimestamp {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	totalNode := consensus.Decider.ParticipantsCount()
+	// diff is at least 1, and it won't exceeded the totalNode
+	diff := uint64(((curTimestamp - blockTimestamp) / viewChangeTimeout) % int64(totalNode))
+	if diff == 0 {
+		diff = 1
+	}
+	nextViewID := diff + consensus.current.GetCurBlockViewID()
+
+	consensus.getLogger().Info().
+		Int64("curTimestamp", curTimestamp).
+		Int64("blockTimestamp", blockTimestamp).
+		Uint64("nextViewID", nextViewID).
+		Uint64("curViewID", consensus.current.GetCurBlockViewID()).
+		Msg("[getNextViewID]")
+
+	// duration is always the fixed view change duration for synchronous view change
+	return nextViewID, viewChangeDuration
+}
+
+// getNextLeaderKey uniquely determine who is the leader for given viewID
+// It reads the current leader's pubkey based on the blockchain data and returns
+// the next leader based on the gap of the viewID of the view change and the last
+// know view id of the block.
+func (consensus *Consensus) getNextLeaderKey(viewID uint64) *bls.PublicKeyWrapper {
 	gap := 1
+
+	if viewID > consensus.GetCurBlockViewID() {
+		gap = int(viewID - consensus.GetCurBlockViewID())
+	}
+	var lastLeaderPubKey *bls.PublicKeyWrapper
+	var err error
+	if consensus.ChainReader == nil {
+		consensus.getLogger().Error().Msg("[getNextLeaderKey] ChainReader is nil. Use consensus.LeaderPubKey")
+		lastLeaderPubKey = consensus.LeaderPubKey
+	} else {
+		curHeader := consensus.ChainReader.CurrentHeader()
+		if curHeader == nil {
+			consensus.getLogger().Error().Msg("[getNextLeaderKey] Failed to get current header from blockchain")
+			lastLeaderPubKey = consensus.LeaderPubKey
+		} else {
+			// this is the truth of the leader based on blockchain blocks
+			lastLeaderPubKey, err = consensus.getLeaderPubKeyFromCoinbase(curHeader)
+			if err != nil || lastLeaderPubKey == nil {
+				consensus.getLogger().Error().Err(err).
+					Msg("[getNextLeaderKey] Unable to get leaderPubKey from coinbase. Set it to consensus.LeaderPubKey")
+				lastLeaderPubKey = consensus.LeaderPubKey
```

Comment:
consensus.LeaderPubKey should take priority than consensus.getLeaderPubKeyFromCoinbase(curHeader), because it's the real current leader. If consensus.LeaderPubKey != nil, then we should just use it. If it's nil, we can fall back on the leader of the last confirmed blocks (which could be the same as consensus.LeaderPubKey, but it could be not, ie. previous leader)

Reply:
why consensus.LeaderPubKey is higher priority than the leader from leader from block? I thought the leader from the block is the truth.  Also, I used the leader from the block is to make sure we can always get to the same next leader.

Reply:
because the leader could change automatically, in cases like epoch change, the leader is reset to the first one. Once we have resharding, the leader will change every N blocks automatically. So the last block's leader may not be the leader for the new block (current in consensus).

Reply:
I see your point. However, the algorithm we used is to determine who is the next leader. It is an deterministic algorithm to use the known last leader to determine who is the next leader.  The leaderpubkey is also an in memory variable, which is not deterministic. So, we should still use the true last known leader to determine the next leader.

In case leader changed automatically, the leader's key will be written to blockchain after a block is confirmed.  Noted, automatically leader change should be independent of the view change.

Reply:
The current GetNextLeaderKey() depends on the consensus.LeaderPubKey, which was determined and set by other logics such as the epoch change or view change. In this case, we have to rely on the in-memory data to determine the next leaders and consensus.LeaderPubKey should take priority.

Think about this case, the leader is latest confirmed block is A. A should produce next block, but it didn't. View change started, and after a few rounds, successfully done on leader D. Now, every nodes set their consensus.LeaderPubKey = D. But D could fail to produce the block too. And new view change needs to happen. Note here, there is no new block confirmed, so the latest confirmed block still have leader A. Then if view change start again, the whole process will happen again, and D could fail everything again.

So bottomline is, we have to rely on in-memory state in this case. And it's ok, as long as most of the nodes don't crash at the same time repeatedly

Reply:
If D is failed to produce block, a new view change will start, the view ID will still be based on A, but noted, the time has moved, so D won't be the leader anymore.  The next leader, assuming E, will be the new leader.